src/abstractdataframe/reshape.jl

```diff
@@ -121,137 +198,240 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
+│ 5   │ 5     │ 3     │ 2     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ 2     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value)
+6×3 DataFrame
+│ Row │ id    │ c         │ d         │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, [:id, :a], :variable, :value)
+6×4 DataFrame
+│ Row │ id    │ a     │ c         │ d         │
+│     │ Int64 │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ 1     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 2     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 2     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 3     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value, renamecols=x->Symbol(:_, x))
+6×3 DataFrame
+│ Row │ id    │ _c        │ _d        │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
 ```
 Note that there are some differences between the widened results above.
 """
 function unstack(df::AbstractDataFrame, rowkey::ColumnIndex, colkey::ColumnIndex,
-                 value::ColumnIndex; renamecols::Function=identity)
-    refkeycol = categorical(df[!, rowkey])
-    droplevels!(refkeycol)
-    keycol = categorical(df[!, colkey])
-    droplevels!(keycol)
+                 value::ColumnIndex; renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
+    refkeycol = df[!, rowkey]
+    keycol = df[!, colkey]
     valuecol = df[!, value]
     return _unstack(df, index(df)[rowkey], index(df)[colkey],
-                    keycol, valuecol, refkeycol, renamecols)
+                    keycol, valuecol, refkeycol, renamecols, allowmissing, allowduplicates)
+end
+
+function _unstack_preprocess_vector(v::AbstractVector)
+    v_unique = unique(v)
+    had_missing = any(ismissing, v_unique)
+    v_unique = intersect(levels(v), v_unique)
+    had_missing && (v_unique = vcat(v_unique, [missing]))
+    len_v = length(v_unique)
+    v_map = Dict([x => i for (i,x) in enumerate(v_unique)])
+    # both unique and Dict should use isequal to test for identity of values
+    @assert length(v_map) == length(v_unique)
+    # if there are no missings in v then set reference index of missing to 0
+    col = similar(v, length(v_unique))
+    copyto!(col, v_unique)
+    return col, v_map, get(v_map, missing, 0)
```

Comment:
This assumes `col` has 1-based indexing. Maybe worth `@assert`ing it? Or is this assumption too widespread in the codebase to be worth it?

Reply:
We use `intersect` above, which internally uses `filter` which normally always returns a `Vector` in this case (though it is not documented this way). E.g. OffsetArrays.jl works this way:
```
julia> OA = OffsetArray([1,2,3], -1:1)
3-element OffsetArray(::Array{Int64,1}, -1:1) with eltype Int64 with indices -1:1:
 1
 2
 3

julia> intersect(OA, OA)
3-element Array{Int64,1}:
 1
 2
 3
```
but I will add an assertion as it does not hurt to have it